ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/answers/activity-composite-answer/activityCompositeAnswer.component.ts

```diff
@@ -110,19 +133,25 @@ export class ActivityCompositeAnswer implements OnChanges {
     }
 
     public updateValue(row: number, column: number, value: AnswerValue): void {
-        this.childQuestionBlocks[row][column].answer = value;
+        const currentBlock = this.childQuestionBlocks[row][column];
+
+        currentBlock.answer = value;
         const childAnswers = this.buildComponentAnswers();
         this.block.setAnswer(childAnswers, false);
+
         // No point in emitting if the value is not valid. Not gonna patch it anyways
-        if (this.childQuestionBlocks[row][column].validate()) {
+        if (currentBlock.validate()) {
             const compositeAnswerValue: any[][] = this.childQuestionBlocks.map(childQuestionBlockRow =>
-                childQuestionBlockRow.map((childQuestionBlock) => {
-                    if (childQuestionBlock.validate()) {
-                        return this.buildChildAnswer(childQuestionBlock);
-                    } else {
-                        return this.buildChildAnswer(childQuestionBlock, null);
-                    }
-                }));
+                childQuestionBlockRow
+                    // we don't patch an equation question answer because it is read-only
+                    .filter(childQuestionBlock => !(childQuestionBlock instanceof ActivityEquationQuestionBlock))
```

Comment:
I BELIEVE you can safely remove this line as I don't expect the EquationComponent will emit an answer.

Reply:
Ah. OK!This something that was not really specified clearly in the API, but what you are doing is reasonable.
You have been able to verify this against the Basil example?

I still would like to get rid of the `instanceof` as it tends to bind the code to specific implementation of QuestionBlock.
The only kind of block that this code should be aware is either a generic ActivityBlock or QuestionBlock.
Would suggest then that we extend ActivityQuestionBlock to include a method like `generatesAnswers(): boolean` or something like that.
The default value can be true, but for EquationBlock it returns false. Then we can get rid of `instanceof`

Reply:
1) yep, I verified the behavior on basil dev env (`Equation activity`).
2) so, you mean that check for a questionBlock.questionType does not make sense as well, right ? 
3) `generatesAnswers` means that only EquationQuestion does not change its answer because it is read-only. Is that the idea ?

Reply:
Yeah. That's the idea. Essentially instead of checking whether a block is instanceof something we can just call a method that answers whether object can or cannot do something. At this time, EquationQuestionBlock (or whatever it is called) would be the only QuestionBlock that would answer the generatesAnswers question with a false.

Reply:
4.
 > get rid of the instanceof as it tends to bind the code to specific implementation of QuestionBlock.

I see your point here. 
However seems to me, in this particular case **maybe** we have to bind the code to `ActivityEquationQuestionBlock` (as a specific implementation of `QuestionBlock`). Because we have some special requirements for that type of questions. 
Like, don't patch the equation answer (both itself and being a composite child). 
Also we should set position indexes in a composite answer (for Equation only), but it doesn't connect to the fact that it can/can't generate answers. It's just because the Equation has quite "unnormal" behavior, we were not able to align it with other questions behavior on back-end. Thus we need to tweak it on the client/Angular side.

Reply:
Yes. Perfectly fine to subclass. It is desirable in this case.
What we want to avoid is have ActivityCompositeAnswer be aware of QuestionBlock specializations or subclasses, like `ActivityEquationQuestionBlock`. Adding the method `generatesAnswers()` to QuestionBlock allows us to not have to check the specific type of the Block. We just ask it whether it can do a thing.

Reply:
Fixed in e3738dc34bf88d67c2585152262921740f6ab8ae